src/Lecture2.hs

```diff
@@ -164,7 +186,46 @@ data Knight = Knight
     , knightEndurance :: Int
     }
 
-dragonFight = error "TODO"
+data Chest a = Chest
+  {
+    chestGold :: Int
+    ,chestTreasure :: a
+  }
+data Dragon a = Dragon
+  {
+    dragonColor :: String ,
+    dragonHealth :: Int,
+    dragonFirePower :: Int,
+    dragonExpPoints :: Int,
+    dragonChest :: Chest a
+  }
```

Comment:
That's a great idea! And almost the perfect solution 🔝 

There's one problem with this particular data type definition:

1. It's possible to have a treasure inside green dragon:

```haskell
myDragon = Dragon
  {
    dragonColor = "Green" ,
    dragonHealth = 150,
    dragonFirePower = 70,
    dragonExpPoints = 500,
    dragonChest = Chest 100 Armor
  }
```

Usually it's possible to model types in such a way to avoid the ability to construct invalid values entirely!

Reply:
@chshersh , I couldn't figure this part out still Dmitri, How can i restrict constructors so that everytime the dragon color is green , it will only accept the dragon chest to be the chest with gold , am i looking at the problem the right way ?!